### PR TITLE
[cxx-interop] Import ObjCPropertyDecl of type NS_OPTIONS fields as struct type

### PR DESCRIPTION
**Pull request by plotfi**
Try importing ObjCPropertyDecl field types the C++-Interop-NS_OPTIONS way. This will fix cases such as:

```swift
import UIKit

func f(gesture: UISwipeGestureRecognizer,
       direction: UISwipeGestureRecognizer.Direction) {
  gesture.direction = direction // error
}
```

because it will make sure the field inside class UIGestureRecognizer is of the enum-struct type and not the typedef-rawValue type when importing an ObjC class.

### CODE REVIEW

**Comment by plotfi** [comment]
@swift-ci please test

**Comment by plotfi** [comment]
@swift-ci please test

**Comment by plotfi** [comment]
@swift-ci please test

File: lib/ClangImporter/ImportDecl.cpp [hunk 1]
```diff
@@ -816,6 +816,31 @@ static bool isPrintLikeMethod(DeclName name, const DeclContext *dc) {
 using MirroredMethodEntry =
   std::tuple<const clang::ObjCMethodDecl*, ProtocolDecl*, bool /*isAsync*/>;
 
+ImportedType tryImportOptionsTypeForField(const clang::QualType type,
+                                          ClangImporter::Implementation &Impl) {
+  ImportedType importedType;
+  auto fieldType = type;
```
Comment (zoecarver) [inline lib/ClangImporter/ImportDecl.cpp:822]: Nit: could we just drop const from the parameter?

Comment (plotfi) [reply lib/ClangImporter/ImportDecl.cpp:822]: Sure thing! If you have any other comments please do post them. I will try and post a pr to 5.9 soon.

File: lib/ClangImporter/ImportDecl.cpp [hunk 2]
```diff
@@ -816,6 +816,31 @@ static bool isPrintLikeMethod(DeclName name, const DeclContext *dc) {
 using MirroredMethodEntry =
   std::tuple<const clang::ObjCMethodDecl*, ProtocolDecl*, bool /*isAsync*/>;
 
+ImportedType tryImportOptionsTypeForField(const clang::QualType type,
```
Comment (zoecarver) [inline lib/ClangImporter/ImportDecl.cpp:819]: I know we use this pattern elsewhere, I thought we already had a helper function for this. Could we just update that? 

If not, I don't think we need to say "for field". Also, maybe we can match a similar function and say "as option set type". So maybe `findOptionSetType` or `findOptionSetTypeForTypedef`.

File: lib/ClangImporter/ImportType.cpp [hunk 3]
```diff
@@ -3247,9 +3250,13 @@ ImportedType ClangImporter::Implementation::importAccessorParamsAndReturnType(
   DeclContext *origDC = importDeclContextOf(property,
                                             property->getDeclContext());
   assert(origDC);
+  auto fieldType = isGetter ? clangDecl->getReturnType() : clangDecl->getParamDecl(0)->getType();
+  ImportedType importedNSOptionsType = tryImportOptionsTypeForField(fieldType, *this);
 
   // Import the property type, independent of what kind of accessor this is.
   auto importedType = importPropertyType(property, isFromSystemModule);
```
Comment (zoecarver) [inline lib/ClangImporter/ImportType.cpp:3257]: We don't want to import the type if we have already found something with `tryImportOptionsTypeForField`

File: lib/ClangImporter/ImportType.cpp [hunk 4]
```diff
@@ -3247,9 +3250,13 @@ ImportedType ClangImporter::Implementation::importAccessorParamsAndReturnType(
   DeclContext *origDC = importDeclContextOf(property,
                                             property->getDeclContext());
   assert(origDC);
+  auto fieldType = isGetter ? clangDecl->getReturnType() : clangDecl->getParamDecl(0)->getType();
```
Comment (zoecarver) [inline lib/ClangImporter/ImportType.cpp:3253]: Can you add a test case for when this is not a getter?

File: lib/ClangImporter/ImportType.cpp [hunk 5]
```diff
@@ -3224,6 +3224,9 @@ ImportedType ClangImporter::Implementation::importMethodParamsAndReturnType(
           importedType.isImplicitlyUnwrapped()};
 }
 
+ImportedType tryImportOptionsTypeForField(const clang::QualType type,
+                                          ClangImporter::Implementation &Impl);
+
 ImportedType ClangImporter::Implementation::importAccessorParamsAndReturnType(
```
Comment (zoecarver) [inline lib/ClangImporter/ImportType.cpp:3230]: You don't have to fix it in this patch, but I wonder if we have the same problem when importing a function that takes the `importMethodParamsAndReturnType` path. Can you try using `SWIFT_COMPUTED_PROPERTY` or `SWIFT_NAME(getter:)` to see if it takes the other path/needs an update?

File: lib/ClangImporter/ImportType.cpp (hunk 4 above)
Comment (plotfi) [reply lib/ClangImporter/ImportType.cpp:3253]: Do you mean when it is *only* a getter? I thought `// CHECK-NEXT:   class func setBar(_ bar: Bar)` was testing for the setter.

File: lib/ClangImporter/ImportType.cpp (hunk 3 above)
Comment (plotfi) [reply lib/ClangImporter/ImportType.cpp:3257]: Ah yeah good point. I think I was trying to leave the subsequent parameter-list creation code in tact here, but I see what you mean.

File: lib/ClangImporter/ImportDecl.cpp (hunk 2 above)
Comment (plotfi) [reply lib/ClangImporter/ImportDecl.cpp:819]: I don't think we have a helper function, I grepped everywhere  in ClangImporter that we use `findAnonymousEnumForTypedef` to try and find it but I dont see one. I can make all the other locations that follow the pattern use this helper and drop the `ForField` though.

File: lib/ClangImporter/ImportType.cpp (hunk 5 above)
Comment (plotfi) [reply lib/ClangImporter/ImportType.cpp:3230]: @zoecarver Do you have any specific frameworks like UIKit or CoreGraphics etc in mind I could use to look for one of these?

**Pull request merged**